File: elfsterclone.js
const lastYearsMatches = {
	rachel: "jeri",
  glenn: "dave",
  jeri: "lauren",
  dave: "rachel",
  lauren: "glenn"
}; // last years matches
const players = ['rachel', 'glenn', 'jeri', 'dave', 'lauren']; // new player inputs
const shuffled = shuffle(players, lastYearsMatches, 0);
console.log({shuffled});

// lastYearMatches is optional
function shuffle (players, lastYearsMatches = [], counter) {
	counter++; // debug for number of iterations
  const shuffled = _.shuffle(players); // shuffles players list using Fisher-Yates shuffle
  const combined = _.zipObject(players, shuffled); // combines two arrays into key: match pairs

  // .every returns a single truthy value for every iteration
	const passMatchRequirements = _.every(combined, function(key, value) {
    const notSelfMatch = key !== value; // check we didn't match ourself
    const notLastYearMatch = lastYearsMatches[key] !== value; // check we didn't match last year

    return notSelfMatch && notLastYearMatch;
  });

  if (passMatchRequirements) {
  	console.log('final iterations: ' + counter); // debug
  	return combined; // return our list new verified matches
  } else {
  	return shuffle(players, lastYearsMatches, counter); // recursivly call this function if we didn't pass .every until we do
  }
}

File: elfsterclone.py
import random
 
def generateNewDraw(buyers, noOneGetsSamePersonAsLastYear, lastYearsDraw):
   global shuffleCount
   receivers = []
   for i in range(len(buyers)):
      receivers.append(buyers[i])
   random.shuffle(receivers)
   shuffleCount += 1
   newDraw = []
   for i in range(len(buyers)):
      if buyers[i] == receivers[i] or (noOneGetsSamePersonAsLastYear and buyers[i] in lastYearsDraw and receivers[i] == lastYearsDraw[buyers[i]]):
         if (i < len(buyers) - 1):
            if buyers[i] == receivers[i + 1]:
               return generateNewDraw(buyers, noOneGetsSamePersonAsLastYear, lastYearsDraw)
            if noOneGetsSamePersonAsLastYear and buyers[i] in lastYearsDraw and receivers[i+1] == lastYearsDraw[buyers[i]]:
               return generateNewDraw(buyers, noOneGetsSamePersonAsLastYear, lastYearsDraw)
            temp = receivers[i]
            receivers[i] = receivers[i + 1]
            receivers[i + 1] = temp
         else:
            return generateNewDraw(buyers, noOneGetsSamePersonAsLastYear, lastYearsDraw)   
      newDraw.append({buyers[i]: receivers[i]})
   return newDraw


lastYearsDraw = {"Rachel": "Jeri",
                "Glenn": "Dave",
                "Jeri": "Lauren",
                "Dave": "Rachel",
                "Lauren": "Glenn"}
noOneGetsSamePersonAsLastYear = True
shuffleCount = 0
buyers = ["Rachel", "Glenn", "Jeri", "Dave", "Lauren"]
random.shuffle(buyers)
newDraw = generateNewDraw(buyers, noOneGetsSamePersonAsLastYear, lastYearsDraw)
results = { "count": shuffleCount, "draw": newDraw }
print results
